parcels/compilation/codecompiler.py

```diff
@@ -0,0 +1,274 @@
+        return self._cppargs
+
+    @property
+    def ldargs(self):
+        return self._ldargs
+
+    @property
+    def incdirs(self):
+        return self._incdirs
+
+    @property
+    def libdirs(self):
+        return self._libdirs
+
+    @property
+    def libs(self):
+        return self._libs
+
+    @property
+    def dynlib_ext(self):
+        return self._dynlib_ext
+
+    @property
+    def stclib_ext(self):
+        return self._stclib_ext
+
+    @property
+    def obj_ext(self):
+        return self._obj_ext
+
+    @property
+    def exe_ext(self):
+        return self._exe_ext
+
+
+class GNU_parameters(Compiler_parameters):
+    def __init__(self, cppargs=None, ldargs=None, incdirs=None, libdirs=None, libs=None):
+        super(GNU_parameters, self).__init__()
+        if cppargs is None:
+            cppargs = []
+        if ldargs is None:
+            ldargs = []
+        if incdirs is None:
+            incdirs = []
+        if libdirs is None:
+            libdirs = []
+        if libs is None:
+            libs = []
+
+        Iflags = []
+        if incdirs is not None and isinstance(incdirs, list):
+            for i, dir in enumerate(incdirs):
+                Iflags.append("-I"+dir)
+        Lflags = []
+        if libdirs is not None and isinstance(libdirs, list):
+            for i, dir in enumerate(libdirs):
+                Lflags.append("-L"+dir)
+        lflags = []
+        if libs is not None and isinstance(libs, list):
+            for i, lib in enumerate(libs):
+                lflags.append("-l" + lib)
+
+        cc_env = os.getenv('CC')
+        self._compiler = "mpicc" if MPI else "gcc" if cc_env is None else cc_env
+        opt_flags = ['-g', '-O3']
+        arch_flag = ['-m64' if calcsize("P") == 8 else '-m32']
+        self._cppargs = ['-Wall', '-fPIC']
+        if len(incdirs) > 0:
+            self._cppargs += Iflags
+        self._cppargs += opt_flags + cppargs + arch_flag
+        self._ldargs = ['-shared']
+        if len(Lflags) > 0:
+            self._ldargs += Lflags
+        if len(lflags) > 0:
+            self._ldargs += lflags
+        self._ldargs += ldargs
+        if len(Lflags) > 0:
+            self._ldargs += ['-Wl,-rpath=%s' % (":".join(libdirs))]
+        self._ldargs += arch_flag
+        self._incdirs = incdirs
+        self._libdirs = libdirs
+        self._libs = libs
+        self._dynlib_ext = "so"
+        self._stclib_ext = "a"
+        self._obj_ext = "o"
+        self._exe_ext = ""
+
+
+class Clang_parameters(Compiler_parameters):
+    def __init__(self, cppargs=None, ldargs=None, incdirs=None, libdirs=None, libs=None):
+        super(Clang_parameters, self).__init__()
+        if cppargs is None:
+            cppargs = []
+        if ldargs is None:
+            ldargs = []
+        if incdirs is None:
+            incdirs = []
+        if libdirs is None:
+            libdirs = []
+        if libs is None:
+            libs = []
+        self._compiler = "cc"
```

Comment:
with respect to the lines you marked here, I don't understand your comment. The variable `self_.compiler="cc"` is only there to initialise the variable with _some_ valid value.

Reply:
`_cppargs`, `_ldargs`, `_incdirs`, `_libdirs`, `_libs` could be refactored to a common (super?) initialiser, instead of defining them on each inheriting class

looking at it again, `_cc`, `_dynlib_ext`, `_stclib_ext`, `_obj_ext` would still need to be defined somewhere, and doesn't really make a difference whether this is on the class itself on within the initialiser

Reply:
no, so, those sets of parameters are two different shoes: `_dynlib_ext`, `_stclib_ext` and `_obj_ext` are parameters consistent for a single Operating System - hence they are implicit with the OS-specific class.

The flag parameters are specific to each compiler individually - (a) in term of e.g. if its `-Iflags` or `/Iflags` (-> Microsoft VS Compiler), and (b) thecause if defined on the user level, this compiler class then needs to 'strip' away those includes and library-links that are not on the system, or that cannot be on the system (looking for libm on windows?); (c) the cppargs need to be stripped depending on the compiler's ANSI-C compliance. Due to those difficulties, I had them still in the subclass.

Still, something slightly less verbose could be done, I agree.